### DISCUSSION TITLE
spotless vs ktlint

### DISCUSSION THREAD

**Discussion by umarpro18** [diffplug/spotless#2006] [Q&A]
What are all benefits using spotless when we have ktlint providing auto formatting options as well, like ./gradlew ktlint -f

**Comment by nedtwigg** [comment]
Spotless does these things

- incremental up-to-date
- buildcache
- partial configuration-cache support
  - soon to be full support #1274
- fixes idempotence bugs
- lets you use one formatter to tweak another
  - add license headers with accurate years from git
  - do a regex replace after the formatter to set something how you want
  - `spotless:off/spotless:on` tags for a range even if the formatter you want doesn't support them


All of this might be implemented by the ad-hoc plugin from whatever formatter, we don't track that. But the reason we exist is the formatters have little bugs or people want to tweak them, and all of those bugfixes, workarounds, and tweaks end up here.

**Reply by umarpro18** [reply]
Thanks.
